ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java

```diff
@@ -1634,6 +1637,44 @@ public static ByteBuffer deserializeUserMetadataRecord(CrcInputStream crcStream)
     }
   }
 
+  /**
+   * Creating a {@link ByteBufferInputStream} from the {@link CrcInputStream} by sharing the underlying memory if the
+   * crcStream is built upon a {@link com.github.ambry.commons.CommonUtils.ByteBufferDataInputStream}.
+   * @param crcStream The crcStream to read {@link ByteBuffer} out.
+   * @param dataSize The size of {@link ByteBuffer}.
+   * @return The {@link ByteBufferInputStream}
+   * @throws IOException Unexpected IO errors.
+   */
+  static ByteBufferInputStream getByteBufferInputStreamForBlobRecord(CrcInputStream crcStream, int dataSize)
+      throws IOException {
+    ByteBufferInputStream output = null;
+    InputStream inputStream = crcStream.getUnderlyingInputStream();
+    if (inputStream instanceof CommonUtils.ByteBufferDataInputStream) {
+      ByteBuffer byteBuffer = ((CommonUtils.ByteBufferDataInputStream) inputStream).getBuffer();
+      int startPosition = byteBuffer.position();
+      // exhaust dataSize bytes and update crc for it.
+      byte[] dummy = new byte[4];
```

Comment:
4 is a pretty small size and _might_ impact crc calculation speed. I wonder if increasing it to something like 512 (still small compared to blob size) would help.

Another thing that we could look into would be introducing a method into Crc32 that can operate directly on bytebuffers without a copy. Java's CRC32 class has such a method.

Reply:
The latter appears doable with small changes to the existing code that deals with byte arrays:
```
  public void update(ByteBuffer b) {
    int localCrc = crc;

    while (b.remaining() > 7) {
      final int c0 = (b.get() ^ localCrc) & 0xff;
      final int c1 = (b.get() ^ (localCrc >>>= 8)) & 0xff;
      final int c2 = (b.get() ^ (localCrc >>>= 8)) & 0xff;
      final int c3 = (b.get() ^ (localCrc >>>= 8)) & 0xff;
      localCrc = (T[T8_7_start + c0] ^ T[T8_6_start + c1]) ^ (T[T8_5_start + c2] ^ T[T8_4_start + c3]);

      final int c4 = b.get() & 0xff;
      final int c5 = b.get() & 0xff;
      final int c6 = b.get() & 0xff;
      final int c7 = b.get() & 0xff;

      localCrc ^= (T[T8_3_start + c4] ^ T[T8_2_start + c5]) ^ (T[T8_1_start + c6] ^ T[T8_0_start + c7]);
    }

    /* loop unroll - duff's device style */
    switch (b.remaining()) {
      case 7:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 6:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 5:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 4:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 3:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 2:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      case 1:
        localCrc = (localCrc >>> 8) ^ T[T8_0_start + ((localCrc ^ b.get()) & 0xff)];
      default:
        /* nothing */
    }

    // Publish crc out to object
    crc = localCrc;
  }
```

You would need to add a method to get the `Crc32` method out of the `CrcInputStream` which is arguably a little hacky.

Reply:
this might actually be a good idea given that random access to direct buffer seems to have a edge over heap buffer or just random byte array https://www.javamex.com/tutorials/io/nio_buffer_performance.shtml